rclcpp/src/rclcpp/executor.cpp

```diff
@@ -745,227 +593,133 @@ Executor::execute_client(
 }
 
 void
-Executor::wait_for_work(std::chrono::nanoseconds timeout)
-{
-  TRACETOOLS_TRACEPOINT(rclcpp_executor_wait_for_work, timeout.count());
-  {
-    std::lock_guard<std::mutex> guard(mutex_);
+Executor::collect_entities()
+{
+  // Get the current list of available waitables from the collector.
+  rclcpp::executors::ExecutorEntitiesCollection collection;
+  this->collector_.update_collections();
+  auto callback_groups = this->collector_.get_all_callback_groups();
+  rclcpp::executors::build_entities_collection(callback_groups, collection);
+
+  // Make a copy of notify waitable so we can continue to mutate the original
+  // one outside of the execute loop.
+  // This prevents the collection of guard conditions in the waitable from changing
+  // while we are waiting on it.
+  if (notify_waitable_) {
+    current_notify_waitable_ = std::make_shared<rclcpp::executors::ExecutorNotifyWaitable>(
+      *notify_waitable_);
+    auto notify_waitable = std::static_pointer_cast<rclcpp::Waitable>(current_notify_waitable_);
+    collection.waitables.insert({notify_waitable.get(), {notify_waitable, {}}});
+  }
+
+  // We must remove expired entities here, so that we don't continue to use older entities.
+  // See https://github.com/ros2/rclcpp/issues/2180 for more information.
+  current_collection_.remove_expired_entities();
```

Comment:
I'm interested in this line, what prevents the same problem occurring after this line but before the next ones? Does this function also lock them all to prevent them from expiring?

Reply:
The next section (syncronizing the new collection into the waitset) also additionally re-locks all of the entity weak_ptrs into shared_pts.

This specific line was introduced because the collection/current_collection maps are using the rcl handles as the key and there is a _good chance_ that handles end up getting re-used if you allocate/deallocate them in a tight loop.  The handles will then conflict and the collections won't get updated correctly.

After talking to @clalancette, I think we may be able to minimize some of the iterations over the collections here, but as a follow-on if possible.

My longer term thought would be to include some capabilities in the waitset for determining which entities are already members and allow some of this logic to be moved into the waitset itself, which should tidy these iterations up.

Reply:
> because the collection/current_collection maps are using the rcl handles as the key and there is a good chance that handles end up getting re-used if you allocate/deallocate them in a tight loop.

If they're being used as keys then they should be held as weak or shared pointers, not just raw pointer values, that way you can know if they have expired. I believe the control block associated with a shared pointer and connected weak pointers are separate from the pointer address and will not be subject to the reuse issue you're describing.

Given that, I don't see how this line actually addresses the issue, because like I mentioned above if between this line and the next ones a rcl handle is re-used you have the same problem no? The only way to prevent that is to try and lock the rcl handle, remove it if you cannot lock it, then keep the shared pointer until your done with the next steps. This line and the reasons that led to it are all very suspect to me, unless I misunderstand them still.